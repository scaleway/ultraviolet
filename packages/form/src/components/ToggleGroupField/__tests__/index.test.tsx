import { screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import { renderWithForm } from '@utils/test'
import { describe, expect, test, vi } from 'vitest'
import { ToggleGroupField } from '..'

describe('GroupField', () => {
  test('should render correctly checked', async () => {
    const { asFragment } = renderWithForm(
      <ToggleGroupField onChange={() => {}} name="Group" legend="Label">
        <ToggleGroupField.Toggle
          name="value-1"
          value="value-1"
          label="value 1"
        />
        <ToggleGroupField.Toggle
          name="value-2"
          value="value-2"
          label="value 1"
        />
      </ToggleGroupField>,
      {
        defaultValues: {
          Group: [],
        },
      },
    )
    const [firstInput, secondInput] = screen.getAllByRole('checkbox', {
      hidden: true,
    })
    await userEvent.click(secondInput)

    expect(firstInput).not.toBeChecked()
    expect(secondInput).toBeChecked()
    expect(asFragment()).toMatchSnapshot()
  })

  test('should trigger events correctly with required prop', async () => {
    const onChange = vi.fn(() => {})

    const { asFragment } = renderWithForm(
      <ToggleGroupField
        name="test"
        onChange={onChange}
        legend="ToggleGroupField events"
        required
      >
        <ToggleGroupField.Toggle
          name="value-1"
          value="value-1"
          label="value 1"
        />
        <ToggleGroupField.Toggle
          name="value-2"
          value="value-2"
          label="value 1"
        />
      </ToggleGroupField>,
      {
        defaultValues: {
          test: [],
        },
      },
    )
    const input = screen.getAllByRole('checkbox', { hidden: true })[0]
    await userEvent.click(input)
    expect(onChange).toBeCalledTimes(1)
    expect(input).toBeChecked()

    await userEvent.click(input)
    expect(onChange).toBeCalledTimes(2)
    expect(input).not.toBeChecked()
    expect(asFragment()).toMatchSnapshot()
  })
})
