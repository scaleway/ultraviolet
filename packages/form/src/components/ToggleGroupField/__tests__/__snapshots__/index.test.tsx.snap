// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GroupField > should render correctly checked 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: contents;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-4 {
  border: none;
  padding: 0;
  margin: 0;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.75rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.25rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-10 {
  color: #222638;
  font-size: 1rem;
  font-family: Inter,Asap,sans-serif;
  font-weight: 500;
  letter-spacing: 0;
  line-height: 1.5rem;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 1rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  cursor: pointer;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.emotion-14:active .emotion-23[data-disabled='false']:after {
  width: calc(1rem * 1.3775);
}

.emotion-14[aria-disabled="true"] {
  cursor: not-allowed;
  color: #b5b7bd;
}

.emotion-16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.125rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-18 {
  display: grid;
  grid-template-columns: auto 1fr;
  gap: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
}

.emotion-20 {
  color: #3f4250;
  font-size: 1rem;
  font-family: Inter,Asap,sans-serif;
  font-weight: 400;
  letter-spacing: 0;
  line-height: 1.5rem;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-22 {
  box-sizing: content-box;
  outline: none;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 1rem;
  position: relative;
  -webkit-transition: all 300ms;
  transition: all 300ms;
  background-color: #e9eaeb;
  width: 3rem;
  height: 1.5rem;
}

.emotion-22:hover {
  background-color: #d9dadd;
}

.emotion-22:after {
  content: "";
  position: absolute;
  top: calc(1.5rem / 2 - 1rem / 2);
  left: 5px;
  width: 1rem;
  height: 1rem;
  border-radius: 100%;
  background-color: #ffffff;
  -webkit-transition: all 300ms;
  transition: all 300ms;
}

.emotion-22:focus-within,
.emotion-22:focus {
  box-shadow: 0px 0px 0px 3px #151a2d5c;
}

.emotion-22[data-disabled="false"]:active:after {
  width: calc(1rem * 1.3775);
}

.emotion-22[data-checked="true"] {
  color: #222638;
  background-color: #8c40ef;
}

.emotion-22[data-checked="true"]:hover {
  background-color: #792dd4;
}

.emotion-22[data-checked="true"]:after {
  left: calc(100% - 5px);
  -webkit-transform: translateX(-100%);
  -moz-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
}

.emotion-22[data-checked="true"]:focus-within,
.emotion-22[data-checked="true"]:focus {
  box-shadow: 0px 0px 0px 3px #8c40ef40;
}

.emotion-22[data-disabled="true"] {
  background: #f3f3f4;
}

.emotion-22[data-disabled="true"][data-checked="true"] {
  background: #e5dbfd;
}

.emotion-22[data-error="true"] {
  background-color: #ffebf2;
}

.emotion-22[data-error="true"]:focus-within,
.emotion-22[data-error="true"]:focus {
  box-shadow: 0px 0px 0px 3px #f91b6c40;
}

.emotion-22[data-error="true"][data-checked="true"] {
  background-color: #e51963;
}

.emotion-22[data-error="true"][data-disabled="true"] {
  background-color: #fff3f7;
}

.emotion-22[data-error="true"][data-disabled="true"][data-checked="true"] {
  background-color: #ffd3e3;
}

.emotion-24 {
  position: absolute;
  opacity: 0;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  cursor: pointer;
}

.emotion-24[disabled] {
  cursor: not-allowed;
}

<div
    data-testid="testing"
  >
    <form
      class="emotion-0 emotion-1"
      novalidate=""
    >
      <div
        class="emotion-2 emotion-3"
      >
        <fieldset
          class="emotion-4 emotion-5"
        >
          <div
            class="emotion-6 emotion-3"
          >
            <div
              class="emotion-8 emotion-3"
            >
              <legend
                class="emotion-10 emotion-11"
              >
                Label 
              </legend>
            </div>
            <div
              class="emotion-12 emotion-3"
            >
              <label
                aria-disabled="false"
                class="emotion-14 emotion-15"
              >
                <div
                  class="emotion-16 emotion-3"
                >
                  <div
                    class="emotion-18 emotion-19"
                  >
                    <span
                      class="emotion-20 emotion-11"
                    >
                      value 1
                    </span>
                  </div>
                </div>
                <div
                  class="emotion-22 emotion-23"
                  data-checked="false"
                  data-disabled="false"
                  data-error="false"
                >
                  <input
                    aria-checked="false"
                    class="emotion-24 emotion-25"
                    id="«r1»"
                    name="Group.value-1"
                    type="checkbox"
                    value="value-1"
                  />
                </div>
              </label>
              <label
                aria-disabled="false"
                class="emotion-14 emotion-15"
              >
                <div
                  class="emotion-16 emotion-3"
                >
                  <div
                    class="emotion-18 emotion-19"
                  >
                    <span
                      class="emotion-20 emotion-11"
                    >
                      value 1
                    </span>
                  </div>
                </div>
                <div
                  class="emotion-22 emotion-23"
                  data-checked="true"
                  data-disabled="false"
                  data-error="false"
                >
                  <input
                    aria-checked="true"
                    class="emotion-24 emotion-25"
                    id="«r4»"
                    name="Group.value-2"
                    type="checkbox"
                    value="value-2"
                  />
                </div>
              </label>
            </div>
          </div>
        </fieldset>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`GroupField > should trigger events correctly with required prop 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: contents;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-4 {
  border: none;
  padding: 0;
  margin: 0;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.75rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.25rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-10 {
  color: #222638;
  font-size: 1rem;
  font-family: Inter,Asap,sans-serif;
  font-weight: 500;
  letter-spacing: 0;
  line-height: 1.5rem;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-13 {
  vertical-align: middle;
  fill: #b3144d;
  height: 8px;
  width: 8px;
  min-width: 8px;
  min-height: 8px;
  vertical-align: super;
}

.emotion-13 .fillStroke {
  stroke: #b3144d;
  fill: none;
}

.emotion-15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 1rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  cursor: pointer;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.emotion-17:active .emotion-26[data-disabled='false']:after {
  width: calc(1rem * 1.3775);
}

.emotion-17[aria-disabled="true"] {
  cursor: not-allowed;
  color: #b5b7bd;
}

.emotion-19 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.125rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-21 {
  display: grid;
  grid-template-columns: auto 1fr;
  gap: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
}

.emotion-23 {
  color: #3f4250;
  font-size: 1rem;
  font-family: Inter,Asap,sans-serif;
  font-weight: 400;
  letter-spacing: 0;
  line-height: 1.5rem;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-25 {
  box-sizing: content-box;
  outline: none;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 1rem;
  position: relative;
  -webkit-transition: all 300ms;
  transition: all 300ms;
  background-color: #e9eaeb;
  width: 3rem;
  height: 1.5rem;
}

.emotion-25:hover {
  background-color: #d9dadd;
}

.emotion-25:after {
  content: "";
  position: absolute;
  top: calc(1.5rem / 2 - 1rem / 2);
  left: 5px;
  width: 1rem;
  height: 1rem;
  border-radius: 100%;
  background-color: #ffffff;
  -webkit-transition: all 300ms;
  transition: all 300ms;
}

.emotion-25:focus-within,
.emotion-25:focus {
  box-shadow: 0px 0px 0px 3px #151a2d5c;
}

.emotion-25[data-disabled="false"]:active:after {
  width: calc(1rem * 1.3775);
}

.emotion-25[data-checked="true"] {
  color: #222638;
  background-color: #8c40ef;
}

.emotion-25[data-checked="true"]:hover {
  background-color: #792dd4;
}

.emotion-25[data-checked="true"]:after {
  left: calc(100% - 5px);
  -webkit-transform: translateX(-100%);
  -moz-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
}

.emotion-25[data-checked="true"]:focus-within,
.emotion-25[data-checked="true"]:focus {
  box-shadow: 0px 0px 0px 3px #8c40ef40;
}

.emotion-25[data-disabled="true"] {
  background: #f3f3f4;
}

.emotion-25[data-disabled="true"][data-checked="true"] {
  background: #e5dbfd;
}

.emotion-25[data-error="true"] {
  background-color: #ffebf2;
}

.emotion-25[data-error="true"]:focus-within,
.emotion-25[data-error="true"]:focus {
  box-shadow: 0px 0px 0px 3px #f91b6c40;
}

.emotion-25[data-error="true"][data-checked="true"] {
  background-color: #e51963;
}

.emotion-25[data-error="true"][data-disabled="true"] {
  background-color: #fff3f7;
}

.emotion-25[data-error="true"][data-disabled="true"][data-checked="true"] {
  background-color: #ffd3e3;
}

.emotion-27 {
  position: absolute;
  opacity: 0;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  cursor: pointer;
}

.emotion-27[disabled] {
  cursor: not-allowed;
}

<div
    data-testid="testing"
  >
    <form
      class="emotion-0 emotion-1"
      novalidate=""
    >
      <div
        class="emotion-2 emotion-3"
      >
        <fieldset
          class="emotion-4 emotion-5"
        >
          <div
            class="emotion-6 emotion-3"
          >
            <div
              class="emotion-8 emotion-3"
            >
              <legend
                class="emotion-10 emotion-11"
              >
                ToggleGroupField events 
                <svg
                  class="emotion-12 emotion-13 emotion-14"
                  viewBox="0 0 20 20"
                >
                  <path
                    d="M10 2h4l-.79 7.91 6.45-4.64 2 3.46L14.42 12l7.24 3.27-2 3.46-6.45-4.64L14 22h-4l.79-7.91-6.45 4.64-2-3.46L9.58 12 2.34 8.73l2-3.46 6.45 4.64z"
                  />
                </svg>
              </legend>
            </div>
            <div
              class="emotion-15 emotion-3"
            >
              <label
                aria-disabled="false"
                class="emotion-17 emotion-18"
              >
                <div
                  class="emotion-19 emotion-3"
                >
                  <div
                    class="emotion-21 emotion-22"
                  >
                    <span
                      class="emotion-23 emotion-11"
                    >
                      value 1
                    </span>
                  </div>
                </div>
                <div
                  class="emotion-25 emotion-26"
                  data-checked="false"
                  data-disabled="false"
                  data-error="false"
                >
                  <input
                    aria-checked="false"
                    class="emotion-27 emotion-28"
                    id="«r8»"
                    name="test.value-1"
                    type="checkbox"
                    value="value-1"
                  />
                </div>
              </label>
              <label
                aria-disabled="false"
                class="emotion-17 emotion-18"
              >
                <div
                  class="emotion-19 emotion-3"
                >
                  <div
                    class="emotion-21 emotion-22"
                  >
                    <span
                      class="emotion-23 emotion-11"
                    >
                      value 1
                    </span>
                  </div>
                </div>
                <div
                  class="emotion-25 emotion-26"
                  data-checked="false"
                  data-disabled="false"
                  data-error="false"
                >
                  <input
                    aria-checked="false"
                    class="emotion-27 emotion-28"
                    id="«rb»"
                    name="test.value-2"
                    type="checkbox"
                    value="value-2"
                  />
                </div>
              </label>
            </div>
          </div>
        </fieldset>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
